### PR TITLE
[SPARK-28697][SQL] Invalidate Database/Table names starting with underscore

### PR DESCRIPTION
**Pull request by ajithme**
## What changes were proposed in this pull request?

I think we should disallow if a identifier starts with _ for create database and create table
Partially we can see its effect in SPARK-28697 where as the table name starts with _ (like _sampleTable) , the FileFormat assumes it to be a hidden folder and do not list it which causes unusual behavior

## How was this patch tested?

Avoiding creating tables and databases with names starting from underscore. Added test case for same

### CODE REVIEW

**Comment by ajithme** [comment]
@dongjoon-hyun @cloud-fan @HyukjinKwon  please review and let me know your opinion on the fix

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by ajithme** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -102,7 +102,7 @@ class SessionCatalog(
   @GuardedBy("this")
   protected var currentDb: String = formatDatabaseName(DEFAULT_DATABASE)
 
-  private val validNameFormat = "([\\w_]+)".r
+  private val validNameFormat = "([a-zA-Z0-9]+[\\w]*)".r
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: can we use `0-9` as the first letter of table name?

Comment (ajithme) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: yes, same as before.. just restricts name not to start with underscore

**Comment by dongjoon-hyun** [comment]
@ajithme . As you see in the UT failures, this is exactly reverting our previous efforts. Please don't change the UT and try to focus on your issues (Hive tables?).
> SPARK-19059: read file based table whose name starts with underscore

cc @gatorsmile

**Comment by HyukjinKwon** [comment]
Do we need to keep the failed unittest case actually?
Seems Hive doesn't fix it too given HIVE-6431. If that's an unreasonable behaviour, let's drop that support and update migration guide.

**Comment by cloud-fan** [comment]
Wait, does table name starting with `_` work in Spark currently? From SPARK-19059 it seems supported, but from SPARK-28697 it seems not.

**Comment by dilipbiswal** [comment]
@cloud-fan @dongjoon-hyun 
Was just checking the db2 definition of a identifier in [link](https://www.ibm.com/support/knowledgecenter/en/SSEPGG_9.7.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000720.html)

Its defined as following :
```
An ordinary identifier is an uppercase letter followed by zero or more characters, each of which is an uppercase letter, a digit, or the underscore character. Note that lower case letters can be used when specifying an ordinary identifier, but they are converted to uppercase when processed. An ordinary identifier should not be a reserved word.
```

Hive seems to have allowed  digit as first character as well.

```
Identifier
    :
    (Letter | Digit) (Letter | Digit | '_')*
    | {allowQuotedId()}? QuotedIdentifier  /* though at the language level we allow all Identifiers to be QuotedIdentifiers;
                                              at the API level only columns are allowed to be of this form */
    | '`' RegexComponent+ '`'
    ;
```

Not sure why in spark we allowed "_" as starting char to begin with ? Is it match some other system ?

**Comment by maropu** [comment]
Just FYI: pgSQL accepts an underscore for the start of table names and doesn't do a digit;
```
postgres=# create table _t(t int);
CREATE TABLE

postgres=# create table 0t(t int);
ERROR:  syntax error at or near "0"
LINE 1: create table 0t(t int);
```

**Comment by ajithme** [comment]
> Wait, does table name starting with `_` work in Spark currently? From SPARK-19059 it seems supported, but from SPARK-28697 it seems not.

When enable Hive support, spark doesn't support

**Comment by ajithme** [comment]
I think spark allowed "_" to be used as start char for table/database to fall inline with PostGre, but this does deviate from behaviour of Hive.

**Comment by dilipbiswal** [comment]
@maropu @ajithme 
Thanks !! Yeah.. looks like the rule for identifier is different in different systems. I think we started with Presto's grammar. Presto also allows "_" as start char and so does SQL server.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 1 above)
Comment (pvk2727) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: we have restricted only _ by this change ?

Comment (pvk2727) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: and abiding by hive standard

**Review by pvk2727 — COMMENTED** [review]
changes seems fine, do wee need to add positive test cases such as test valid names, to assure the valid format "([a-zA-Z0-9]+[\\w]*)".r

**Comment by cloud-fan** [comment]
According to the discussions in this PR, seems there are many other databases allowing the identifiers to start with `_`. Spark can support it as well without hive support. Shall we add the check only in `HiveExternalCatalog.createTable`?

**Comment by ajithme** [comment]
> According to the discussions in this PR, seems there are many other databases allowing the identifiers to start with `_`. Spark can support it as well without hive support. Shall we add the check only in `HiveExternalCatalog.createTable`?

@cloud-fan @HyukjinKwon 
This for inputs. Would it not differ spark behavior when hive support is enabled or disabled.? (Hive cannot support _ starting identifiers due to limitations in FileFormat which affect its extended classes too) I think the identifier support inside spark should be uniform. Please correct me if i am wrong. Else i am ok to add check only at ``HiveExternalCatalog.createTable`` so that we can get more proper error message and mark this as a limitation when hive support is enabled

**Pull request closed without merging**